File: test/test_engine.py
"""Python test: engine reproduces the anchor scores captured from the official
IRCC calculator, and agrees with the JS engine on a batch of random profiles.
Run:  python test/test_engine.py
"""
import os, sys, json, random, subprocess

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
sys.path.insert(0, os.path.join(ROOT, "engine"))
import crs_engine as E

def L(v): return {"q5i-b-" + a: v for a in E.ABIL}

ANCHORS = [
    (dict(q1="F", q3="M", q4="E", **{"q5i-a": "B"}, **L("G")), 379),
    (dict(q1="F", q3="M", q4="E", **{"q5i-a": "B"}, **L("H")), 391),
    (dict(q1="E", q2i="A", q2ii="B", q3="M", q4="E", **{"q5i-a": "B"}, **L("G")), 353),
    (dict(q1="F", q3="M", q4="E", q9="B", **{"q5i-a": "B"}, **L("G")), 979),
    (dict(q1="F", q3="M", q4="E", q6i="B", **{"q5i-a": "B"}, **L("G")), 432),
]

def test_anchors():
    for prof, exp in ANCHORS:
        got = E.score(E.full(prof))["GRAND_TOTAL"]
        assert got == exp, "anchor %d != %d" % (got, exp)
    print("Python anchors OK (%d cases)" % len(ANCHORS))

if __name__ == "__main__":
    test_anchors()
    print("All Python tests passed.")

File: test/test_engine.js
/* Node test: the JS engine reproduces the anchor scores captured from the
   official IRCC calculator. Run:  node test/test_engine.js  */
const fs = require("fs");
const path = require("path");
const CRSEngine = require("../engine/crs-engine.js");
const RULES = JSON.parse(fs.readFileSync(path.join(__dirname, "..", "data", "crs-rules.json"), "utf8"));
const ENG = CRSEngine.createEngine(RULES);

const L = (v) => ({ "q5i-b-speaking": v, "q5i-b-listening": v, "q5i-b-reading": v, "q5i-b-writing": v });
const cases = [
  [Object.assign({ q1: "F", q3: "M", q4: "E", "q5i-a": "B" }, L("G")), 379],
  [Object.assign({ q1: "F", q3: "M", q4: "E", "q5i-a": "B" }, L("H")), 391],
  [Object.assign({ q1: "E", q2i: "A", q2ii: "B", q3: "M", q4: "E", "q5i-a": "B" }, L("G")), 353],
  [Object.assign({ q1: "F", q3: "M", q4: "E", q9: "B", "q5i-a": "B" }, L("G")), 979],
  [Object.assign({ q1: "F", q3: "M", q4: "E", q6i: "B", "q5i-a": "B" }, L("G")), 432],
];
let fail = 0;
for (const [prof, exp] of cases) {
  const got = ENG.score(ENG.full(prof)).GRAND_TOTAL;
  if (got !== exp) { console.error("FAIL", exp, "got", got); fail++; }
}
if (fail) { console.error(fail + " anchor(s) failed"); process.exit(1); }
console.log("JS engine anchors OK (" + cases.length + " cases)");
